## 将字符串转化成小写字母

实现函数 ToLowerCase()，该函数接收一个字符串参数 str，并将该字符串中的大写字母转换成小写字母，之后返回新的字符串。

In [ ]:
def toLowerCase(str: str):
    #Put you anwser here
    return str.lower()

# 输出内容
A = 'HELLO'
a = toLowerCase(A)
print(f"转换之前{A}")
print(f"转换之后{a}")

转换之前HELLO
转换之后hello


## 反转字符串中的单词

给定一个字符串，你需要反转字符串中每个单词的字符顺序，同时仍保留空格和单词的初始顺序。

示例 1:

> 输入: "Let's take deep learning contest"

> 输出: "s'teL ekat peed gninrael tsetnoc"



In [ ]:
def reverseWords(s: str):
    #Put you anwser here
    words = s.split()
    reverse_word = [word[::-1] for word in words]
    str = ''.join(reverse_word)
    return str

# 输出内容
input = "Let's take deep learning contest"
output = reverseWords(input)
print(f"反转之前: {input}")
print(f"反转之后: {output}")

反转之前: Let's take deep learning contest
反转之后: s'teLekatpeedgninraeltsetnoc


## 排列组合

有四个数字：1、2、3、4，能组成多少个互不相同且无重复数字的三位数？各是多少？

In [ ]:
def Permutation(input_num):
    list = []
    #Code goes here
    # 如range(3)等价于range(0,3),表示创建0-3的数组
    for i in range(len(input_num)):   
        for j in range(len(input_num)):
            if i == j:    # 如果两个数相等则跳过
                continue  
            for k in range(len(input_num)):
                if k == i or k == j:
                    continue
                list.append(str(input_num[i])+str(input_num[j])+str(input_num[k]))
    return list

# 输出内容
input = [1, 2, 3, 4]
output = Permutation(input)
out_len = len(output)
print(f"输入数据: {input}")
print(f"排列组合数量: {out_len}")
print(f"排列组合分别为: {output}")

输入数据: [1, 2, 3, 4]
排列组合数量: 24
排列组合分别为: ['123', '124', '132', '134', '142', '143', '213', '214', '231', '234', '241', '243', '312', '314', '321', '324', '341', '342', '412', '413', '421', '423', '431', '432']


## 阶乘
让函数FirstFactorial（num）接受传递的num参数并返回它的阶乘。例如：如果num = 4，则程序应返回（4 * 3 * 2 * 1）= 24.对于测试用例，范围将介于1和18之间，输入将始终为整数

In [ ]:
def FirstFactorial(num): 
    output = 1
    for i in range(1, num):
        output *= i
    return output
    
# keep this function call here 
input = 7
output = FirstFactorial(int(input))
print(f'{input}的阶乘是{output}')

7的阶乘是720


## 整型转罗马数字

罗马数字 3 写做 III ，即为3个并列的 1。12 写做 XII ，即为 X + II 。 26 写做  XXVI, 即为 XX + V + I 。
也有一些特例, 例如4, 可以放到V的左边, IV表示4. 

要将整数转换为罗马数字，我们需要理解罗马数字的构成规则。罗马数字使用特定的字符表示数字，并且这些字符可以按照一定的规则组合。每个数字有一个对应的罗马数字符号，例如：
1 = I
4 = IV
5 = V
9 = IX
10 = X
40 = XL
50 = L
90 = XC
100 = C
400 = CD
500 = D
900 = CM
1000 = M

In [ ]:
def intToRoman( num: int):
    # 罗马数字的符号与对应的整数值的映射
    roman_numerals = [
        (1000, 'M'), (900, 'CM'), (500, 'D'), (400, 'CD'),
        (100, 'C'), (90, 'XC'), (50, 'L'), (40, 'XL'),
        (10, 'X'), (9, 'IX'), (5, 'V'), (4, 'IV'),
        (1, 'I')
    ]

    roman = ""   # 存放最终的罗马数字结果
    
    # 遍历罗马数字符号和对应的整数值
    for value, symbol in roman_numerals:
        # 计算当前符号应该重复的次数
        while num >= value:
            roman += symbol
            num -= value  # 减去相应的数值

    return roman

print(intToRoman(3))   # 输出 "III"
print(intToRoman(12))  # 输出 "XII"
print(intToRoman(26))  # 输出 "XXVI"
print(intToRoman(4))   # 输出 "IV"
print(intToRoman(1994)) # 输出 "MCMXCIV"   

III
XII
XXVI
IV
MCMXCIV


## 盛水最多的容器

给定 n 个非负整数 a1，a2，...，an，每个数代表坐标中的一个点 (i, ai) 。在坐标内画 n 条垂直线，垂直线 i 的两个端点分别为 (i, ai) 和 (i, 0)。找出其中的两条线，使得它们与 x 轴共同构成的容器可以容纳最多的水。

说明：你不能倾斜容器，且 n 的值至少为 2。

![](https://ai-studio-static-online.cdn.bcebos.com/8201601c528c46e8af4bab4e65bd8cc4973d1c44f1a24b0da4238fc77a650117)

图中垂直线代表输入数组 [1,8,6,2,5,4,8,3,7]。在此情况下，容器能够容纳水（表示为蓝色部分）的最大值为 49。

示例:
```
输入: [1,8,6,2,5,4,8,3,7]
输出: 49
```

In [ ]:
from typing import List
def maxArea(height: List[int]):
    left, right = 0, len(height)-1
    max_area = 0
    # code goes here 
    while left <right:
        # 计算当前容器的容量
        area = min(height[left], height[right]) * (right - left)
        # 更新最大容器容量
        max_area = max(max_area, area)

        # 移动指针
        if height[left] < height[right]:
            left += 1
        else:
            right -= 1

    return max_area


input_num = [1,8,6,2,5,4,8,3,7]
print(maxArea(input_num))

49


## 矩阵置零


给定一个 m * n 的矩阵，如果一个元素为 0，则将其所在行和列的所有元素都设为 0。请使用原地算法。

示例 1:
```
输入: 
[
  [1,1,1],
  [1,0,1],
  [1,1,1]
]
输出: 
[
  [1,0,1],
  [0,0,0],
  [1,0,1]
]
```
示例 2:
```
输入: 
[
  [0,1,2,0],
  [3,4,5,2],
  [1,3,1,5]
]
输出: 
[
  [0,0,0,0],
  [0,4,5,0],
  [0,3,1,0]
]
```
进阶:

* 一个直接的解决方案是使用  O(mn) 的额外空间，但这并不是一个好的解决方案。
* 一个简单的改进方案是使用 O(m + n) 的额外空间，但这仍然不是最好的解决方案。
* 你能想出一个常数空间的解决方案吗？




In [ ]:
import numpy as np
def setZeroes( matrix):
    # Do not return anything, modify matrix in-place instead.
    # Code goes here
    out_matrix = np.copy(matrix)
    # row_num, col_num = matrix.shape   # 矩阵的行列数
    row_num = len(matrix)       # 矩阵的行数
    col_num = len(matrix[0])    # 矩阵的列数
    # print(row_num, col_num)
    
    zero_col = []       # 记录矩阵中0所在的列
    for i in range(row_num):
        for j in range(col_num):
            # print(matrix)
            if matrix[i][j] == 0:
                out_matrix[i] = [0]*len(matrix[i])
                zero_col.append(j)
    zero_col = set(zero_col)            
    # print(zero_col)

    # 将0所在的列都设置为0
    for n in zero_col:     
        for k in range(row_num):  
            out_matrix[k][n] = 0
    return out_matrix

input_num1 = [
  [0,1,1],
  [1,0,1],
  [1,0,1]
]
print(input_num1)
print(setZeroes(input_num1))

input_num2 = [
  [0,1,2,0],
  [3,4,5,2],
  [1,3,1,5]
]
print(input_num2)
print(setZeroes(input_num2))


[[0, 1, 1], [1, 0, 1], [1, 0, 1]]
[[0 0 0]
 [0 0 0]
 [0 0 0]]
[[0, 1, 2, 0], [3, 4, 5, 2], [1, 3, 1, 5]]
[[0 0 0 0]
 [0 4 5 0]
 [0 3 1 0]]


## 文件操作

有两个磁盘文件A和B,各存放一行字母,要求把这两个文件中的信息合并(按字母顺序排列), 输出到一个新文件C中

In [ ]:
def combination(file1, file2):
    #将字母进行拼接
    file3 = file1 + file2
    print(f"排序前{file3}")
    # 进行排序
    file3 = sorted(file3)   # 字符分割
    file3 = ''.join(file3)  # 将内容进行合并
    return file3


A = "avds"   #{a, v, d, s}
B = "mnai"   #{m, n, a, i}
print(combination(A, B))

排序前avdsmnai
aadimnsv


# 更高难度练习:


### 一、猜数字
经典的猜数字游戏，几乎所有人学编程时都会做。

功能描述：随机选择一个三位以内的数字作为答案。用户输入一个数字，程序会提示大了或是小了，直到用户猜中。

 

### 二、FizzBuzz
另一道经典编程题。

功能描述：遍历并打印0到100，如果数字能被3整除，显示Fizz；如果数字能被5整除，显示Buzz；如果能同时被3和5整除，就显示FizzBuzz。结果应该类似：0,1,2，Fizz，4，Buzz，6……14，FizzBuzz，16……

 

### 三、猜数字的AI
和猜数字一样，不过这次是设计一个能猜数字的AI

功能描述：用户输入一个单位以内的数字，AI要用最少的次数猜中，并且显示出猜的次数和数字。

 

### 四、整点报时
老式挂钟会在整点的报时，响铃的次数和时间相等。我们设计一个在电脑上运行的报时器。

功能描述：运行后，在每一个整点长响一声，半个整点短响两声。实现睡眠模式，晚上十二点到早上六点不响铃。

 

### 五、敏感词审查
一向痛恨网络审查的你成了某网站的审查员，要求审查网络上违反上头规定的名词。

功能描述：要审查的帖子在这个文本文档里，要求将所有的和谐，三个代表，言论自由，64替换为*号。

 

### 六、文字冒险游戏
你想做个清新脱俗的文字游戏，只在命令行的黑框里运行。

功能描述：尽量参考回合制RPG游戏的模式，游戏剧情自编，要有战斗模式，主角有HP属性，战斗和吃东西会对HP有相应的增减效果。

 

### 七、词频统计
你想分析某个英语写作者的写作风格，你想先从他使用词语的频率入手。

功能描述：分析这个文本文档里的词频，按照出现频率由高到低排列结果，不区分大小写，过滤掉标点（可以使用正则表达式）。结果类似ok：234，play：122，funny：78

 

### 八、Markdown转换器
你一向喜欢在Windows的记事本里写文章，为了能方便的生成网页，你使用Markdown的格式写了篇文章，现在需要一个把Markdown转换成HTML格式的程序。

功能描述：将这个文本文档里的内容转换成HTML文档。要尽可能的支持Markdown标签，越多越好。建议使用正则表达式。

 

### 九、邮件轰炸机
你的一个朋友生日快到了，你决定搞个恶作剧——用邮件塞满他的邮箱。

功能描述：首先去注册十个邮箱，然后用这是个邮箱轮流发送随机生成的邮件内容给你的朋友。调整发送的频率，以免被服务器拒绝。

 

### 十、用邮件控制电脑
你的电脑放在宿舍，这时你在教室上课，你又决定搞个恶作剧吓吓装病请假在宿舍玩游戏的室友。

功能描述：使用邮件发送特定的命令，让电脑上的程序判断发信方是否为控制邮箱，如果是就执行发送过来的命令。通过不同的方式实现尽可能多的功能。

 

一、猜数字
经典的猜数字游戏，几乎所有人学编程时都会做。

功能描述：随机选择一个三位以内的数字作为答案。用户输入一个数字，程序会提示大了或是小了，直到用户猜中。

In [ ]:
import random

# 生成一个随机的三位数以内的数字（范围 1 到 999）
answer = random.randint(1, 999)

print("欢迎来到猜数字游戏！")
print("我已经选择了一个 1 到 999 之间的数字，试着猜猜它是多少吧！")

# 进行猜数字的循环
while True:
    # 用户输入数字
    user_guess = int(input("请输入你猜的数字:"))  # 确保没有覆盖 int 函数
    
    # 判断用户输入的数字与答案的关系
    if user_guess < answer:
        print("太小了！再试试吧。")
    elif user_guess > answer:
        print("太大了！再试试吧。")
    else:
        print(f"恭喜你！猜对了，答案就是 {answer}！")
        break  # 猜对了，退出循环


欢迎来到猜数字游戏！
我已经选择了一个 1 到 999 之间的数字，试着猜猜它是多少吧！


TypeError: 'int' object is not callable

二、FizzBuzz
另一道经典编程题。

功能描述：遍历并打印0到100，如果数字能被3整除，显示Fizz；如果数字能被5整除，显示Buzz；如果能同时被3和5整除，就显示FizzBuzz。结果应该类似：0,1,2，Fizz，4，Buzz，6……14，FizzBuzz，16……

In [ ]:
for input_num in range(101):
    if input_num % 3 == 0 and input_num % 5 == 0:
        print("FizzBuzz")
    elif input_num % 3 == 0:
        print("Fizz")    
    elif input_num % 5 == 0:
        print("Buzz")  
    else:
        print(input_num)  

FizzBuzz
1
2
Fizz
4
Buzz
Fizz
7
8
Fizz
Buzz
11
Fizz
13
14
FizzBuzz


三、猜数字的AI
和猜数字一样，不过这次是设计一个能猜数字的AI

功能描述：用户输入一个单位以内的数字，AI要用最少的次数猜中，并且显示出猜的次数和数字。

In [ ]:
import random

def ai_guess_num(min_value, max_value, answer):
    low, high = min_value, max_value
    guess_count = 0
    
    # 开始AI猜数字
    while True:
        guess = (low+high) // 2
        print(f"当前AI猜测的数字是{guess}")
        
        guess_count += 1
        if guess < answer:
            print("输入的数字有点小，请继续猜测！")
            low = guess
        elif guess > answer:
            print("输入的数字有点大，请继续猜测！")
            high = guess
        else:
            print(f"恭喜你猜对了，输入的数据就是{answer}!")
            break
        
    return guess_count


# 生成一个随机的三位数以内的数字（范围 1 到 999）
answer = random.randint(100, 999)
print(answer)
# ai猜数字
guess_num = ai_guess_num(100, 999, answer)
print(f"总共猜测{guess_num}次")


570
当前AI猜测的数字是549
输入的数字有点小，请继续猜测！
当前AI猜测的数字是774
输入的数字有点大，请继续猜测！
当前AI猜测的数字是661
输入的数字有点大，请继续猜测！
当前AI猜测的数字是605
输入的数字有点大，请继续猜测！
当前AI猜测的数字是577
输入的数字有点大，请继续猜测！
当前AI猜测的数字是563
输入的数字有点小，请继续猜测！
当前AI猜测的数字是570
恭喜你猜对了，输入的数据就是570!
总共猜测7次


四、整点报时
老式挂钟会在整点的报时，响铃的次数和时间相等。我们设计一个在电脑上运行的报时器。

功能描述：运行后，在每一个整点长响一声，半个整点短响两声。实现睡眠模式，晚上十二点到早上六点不响铃。

In [ ]:
import time
from datetime import datetime
def sound_bell(count, nn):
    """ 模拟报时，响铃指定次数 """
    for _ in range(count):
        print("叮铃铃！", end=' ')
        time.sleep(0.2*nn)  # 每次铃声之间的间隔

def hourly_chime():
    """ 每小时进行报时和半点报时的功能 """
    while True:
        # 获取当前时间
        current_time = datetime.now()
        hour = current_time.hour
        minute = current_time.minute
        
        # 判断是否处于睡眠时间（12:00 AM 到 6:00 AM）
        if 0 <= hour < 6:
            print(f"当前时间 {current_time.strftime('%H:%M')}，在睡眠模式中，暂不响铃。")
            time.sleep(60)  # 睡眠时间段每次检查一次
            continue
        
        # 判断整点和半点
        if minute == 0:  # 整点
            print(f"当前时间 {current_time.strftime('%H:%M')}，整点报时：")
            sound_bell(hour, 2)  # 整点响铃次数为当前小时数
        elif minute == 30:  # 半点
            print(f"当前时间 {current_time.strftime('%H:%M')}，半点报时：")
            sound_bell(2, 1)  # 半点响铃2次
            
        # 每隔60秒检查一次
        time.sleep(60)

# 运行报时器
if __name__ == "__main__":
    print("报时器启动...")
    hourly_chime()

报时器启动...


KeyboardInterrupt: 

五、敏感词审查
一向痛恨网络审查的你成了某网站的审查员，要求审查网络上违反上头规定的名词。

功能描述：要审查的帖子在这个文本文档里，要求将所有的和谐，三个代表，言论自由，64替换为*号。

In [ ]:
import re
sensitive_words = ['和谐', '三个代表', '言论自由', '64']

# 定义一个函数来处理敏感词替换
def censor_text(text):
    for word in sensitive_words:
        # 使用正则表达式来替换敏感词为*号，保留敏感词长度
        text = re.sub(word, '*' * len(word), text)
    return text

in_sents = "我们现在的社会是和谐美好的，我们应该高举马克思主义中国化旗帜，遵循邓小平理论、三个代表等重要思想。当前社会是提倡自由民主的，我们拥有言论自由、行为自由等，这是社会繁荣昌盛的一个标志。" 
print(censor_text(in_sents))

我们现在的社会是**美好的，我们应该高举马克思主义中国化旗帜，遵循邓小平理论、****等重要思想。当前社会是提倡自由民主的，我们拥有****、行为自由等，这是社会繁荣昌盛的一个标志。


六、文字冒险游戏
你想做个清新脱俗的文字游戏，只在命令行的黑框里运行。

功能描述：尽量参考回合制RPG游戏的模式，游戏剧情自编，要有战斗模式，主角有HP属性，战斗和吃东西会对HP有相应的增减效果。

In [ ]:
import random
import time

# 游戏角色类
class Character:
    def __init__(self, name, hp, attack, defense):
        self.name = name
        self.hp = hp
        self.attack = attack
        self.defense = defense
    
    def is_alive(self):
        return self.hp > 0
    
    def take_damage(self, damage):
        self.hp -= damage
        if self.hp < 0:
            self.hp = 0

    def heal(self, amount):
        self.hp += amount
        print(f"{self.name} 恢复了 {amount} 点生命值！")

    def attack_enemy(self, enemy):
        damage = self.attack - enemy.defense
        if damage < 0:
            damage = 0
        enemy.take_damage(damage)
        print(f"{self.name} 攻击了 {enemy.name}，造成了 {damage} 点伤害！")
    
    def __str__(self):
        return f"{self.name} - HP: {self.hp} 攻击: {self.attack} 防御: {self.defense}"

# 游戏剧情与战斗系统
def start_game():
    # 初始化玩家和敌人
    player = Character("勇者", 100, 20, 5)
    monster = Character("怪物", 50, 15, 3)
    
    print("欢迎来到文字冒险游戏！")
    print("你是一个勇者，面临着可怕的怪物！")
    print(f"你当前的生命值：{player.hp}, 怪物的生命值：{monster.hp}\n")
    
    while player.is_alive() and monster.is_alive():
        # 打印角色信息
        print("\n选择行动：")
        print("1. 攻击怪物")
        print("2. 吃药（恢复20点生命）")
        print("3. 逃跑")
        
        choice = input("请输入你的选择（1/2/3）：")
        
        if choice == "1":
            player.attack_enemy(monster)
            if monster.is_alive():
                print(f"\n怪物反击！")
                monster.attack_enemy(player)
            else:
                print(f"\n{monster.name} 被打败了！")
        
        elif choice == "2":
            player.heal(20)
            monster.attack_enemy(player)
        
        elif choice == "3":
            print("\n你选择了逃跑，游戏结束。")
            break
        
        else:
            print("\n无效的选择，请重新选择！")
            continue
        
        print(f"\n你当前的生命值：{player.hp}, 怪物的生命值：{monster.hp}\n")
        time.sleep(1)

    if player.is_alive():
        print("你赢了！")
    else:
        print("你被怪物打败了。")

# 启动游戏
if __name__ == "__main__":
    start_game()


欢迎来到文字冒险游戏！
你是一个勇者，面临着可怕的怪物！
你当前的生命值：100, 怪物的生命值：50


选择行动：
1. 攻击怪物
2. 吃药（恢复20点生命）
3. 逃跑
请输入你的选择（1/2/3）：勇者 攻击了 怪物，造成了 17 点伤害！

怪物反击！
怪物 攻击了 勇者，造成了 10 点伤害！

你当前的生命值：90, 怪物的生命值：33


选择行动：
1. 攻击怪物
2. 吃药（恢复20点生命）
3. 逃跑
请输入你的选择（1/2/3）：勇者 恢复了 20 点生命值！
怪物 攻击了 勇者，造成了 10 点伤害！

你当前的生命值：100, 怪物的生命值：33


选择行动：
1. 攻击怪物
2. 吃药（恢复20点生命）
3. 逃跑
请输入你的选择（1/2/3）：
你选择了逃跑，游戏结束。
你赢了！


七、词频统计
你想分析某个英语写作者的写作风格，你想先从他使用词语的频率入手。

功能描述：分析这个文本文档里的词频，按照出现频率由高到低排列结果，不区分大小写，过滤掉标点（可以使用正则表达式）。结果类似ok：234，play：122，funny：78

In [ ]:
import re
from collections import Counter
def word_frequency_analysis(texts):
    # 去掉标点符号，只保留单词
    words = re.findall(r'\b\w+\b', texts)
    # 单词频率统计
    word_counts = Counter(words)

    # 按照词频从高到低排序
    sorted_word_counts = word_counts.most_common()

    # 输出结果
    for word, count in sorted_word_counts:
        print(f'{word}: {count}')



in_sents ="""Ok, let’s talk about how things went down yesterday. You remember, right? We were supposed to meet at the park, but you were running a bit late. It’s totally fine, though, we both know how things can get sometimes. I was sitting there for a while, but I didn’t mind. ...
 Actually, it gave me some time to think about stuff. You know, life and all that.""" #
 # When you finally showed up, we had a quick chat about the project we’re working on. Honestly, I think it’s going well so far. We’re both putting in the effort, and that’s all that matters. You’re doing great with your part, and I’ve been trying to make sure we stay on track. It’s not always easy, but we’re managing, and that’s what counts.
# I know there are moments when things feel a bit overwhelming, but that’s normal. I just want you to know that I’m here for you. If you ever need someone to talk to or bounce ideas off of, you can always count on me. We’ve been through a lot together, and I think we’re stronger because of it. So, don’t stress too much. We’ve got this.
#And by the way, I think we should do something fun soon—like grab a coffee or catch a movie. It’ll be a nice break from all the work. What do you think? Sounds good? Ok, well, let me know when you’re free, and we can make it happen."""
# 统计结果
word_frequency_analysis(in_sents)

Counter({'s': 2, 'about': 2, 'how': 2, 'things': 2, 'You': 2, 'were': 2, 'to': 2, 'but': 2, 'a': 2, 'know': 2, 'I': 2, 'Ok': 1, 'let': 1, 'talk': 1, 'went': 1, 'down': 1, 'yesterday': 1, 'remember': 1, 'right': 1, 'We': 1, 'supposed': 1, 'meet': 1, 'at': 1, 'the': 1, 'park': 1, 'you': 1, 'running': 1, 'bit': 1, 'late': 1, 'It': 1, 'totally': 1, 'fine': 1, 'though': 1, 'we': 1, 'both': 1, 'can': 1, 'get': 1, 'sometimes': 1, 'was': 1, 'sitting': 1, 'there': 1, 'for': 1, 'while': 1, 'didn': 1, 't': 1, 'mind': 1, 'Actually': 1, 'it': 1, 'gave': 1, 'me': 1, 'some': 1, 'time': 1, 'think': 1, 'stuff': 1, 'life': 1, 'and': 1, 'all': 1, 'that': 1})
s: 2
about: 2
how: 2
things: 2
You: 2
were: 2
to: 2
but: 2
a: 2
know: 2
I: 2
Ok: 1
let: 1
talk: 1
went: 1
down: 1
yesterday: 1
remember: 1
right: 1
We: 1
supposed: 1
meet: 1
at: 1
the: 1
park: 1
you: 1
running: 1
bit: 1
late: 1
It: 1
totally: 1
fine: 1
though: 1
we: 1
both: 1
can: 1
get: 1
sometimes: 1
was: 1
sitting: 1
there: 1
for: 1
while: 1
didn:

八、Markdown转换器
你一向喜欢在Windows的记事本里写文章，为了能方便的生成网页，你使用Markdown的格式写了篇文章，现在需要一个把Markdown转换成HTML格式的程序。

功能描述：将这个文本文档里的内容转换成HTML文档。要尽可能的支持Markdown标签，越多越好。建议使用正则表达式。

In [ ]:
import re

def markdown_to_html(markdown_text):
    # 处理标题：# 或 ## 或 ### 等
    markdown_text = re.sub(r'^(#)(.*?)$', r'<h1>\2</h1>', markdown_text, flags=re.MULTILINE)  # 处理h1
    markdown_text = re.sub(r'^(##)(.*?)$', r'<h2>\2</h2>', markdown_text, flags=re.MULTILINE)  # 处理h2
    markdown_text = re.sub(r'^(###)(.*?)$', r'<h3>\2</h3>', markdown_text, flags=re.MULTILINE)  # 处理h3
    markdown_text = re.sub(r'^(####)(.*?)$', r'<h4>\2</h4>', markdown_text, flags=re.MULTILINE)  # 处理h4
    markdown_text = re.sub(r'^(#####)(.*?)$', r'<h5>\2</h5>', markdown_text, flags=re.MULTILINE)  # 处理h5
    markdown_text = re.sub(r'^(######)(.*?)$', r'<h6>\2</h6>', markdown_text, flags=re.MULTILINE)  # 处理h6

    # 处理加粗：**加粗文本** 或 __加粗文本__
    markdown_text = re.sub(r'(\*\*|__)(.*?)\1', r'<strong>\2</strong>', markdown_text)

    # 处理斜体：*斜体文本* 或 _斜体文本_
    markdown_text = re.sub(r'(\*|_)(.*?)\1', r'<em>\2</em>', markdown_text)

    # 处理无序列表：- 或 * 或 + 开头
    markdown_text = re.sub(r'^[\*\+\-]\s+(.*)', r'<ul><li>\1</li></ul>', markdown_text, flags=re.MULTILINE)
    
    # 处理有序列表：数字. 开头
    markdown_text = re.sub(r'^\d+\.\s+(.*)', r'<ol><li>\1</li></ol>', markdown_text, flags=re.MULTILINE)

    # 处理链接：[链接文本](url)
    markdown_text = re.sub(r'\[([^\]]+)\]\(([^)]+)\)', r'<a href="\2">\1</a>', markdown_text)

    # 处理图片：![图片文本](url)
    markdown_text = re.sub(r'!\[([^\]]*)\]\(([^)]+)\)', r'<img src="\2" alt="\1">', markdown_text)

    # 处理代码：`代码`
    markdown_text = re.sub(r'`([^`]+)`', r'<code>\1</code>', markdown_text)

    # 处理块级代码：```代码块```
    markdown_text = re.sub(r'```(.*?)```', r'<pre><code>\1</code></pre>', markdown_text, flags=re.DOTALL)

    # 处理段落换行：两个空格后换行
    markdown_text = re.sub(r'  \n', r'<br>', markdown_text)

    # 处理普通文本的段落（这里默认每一行的文本都按段落处理）
    markdown_text = re.sub(r'([^\n]+)', r'<p>\1</p>', markdown_text)

    # 返回转换后的HTML文本
    return markdown_text

# 示例：将Markdown文本转换为HTML
if __name__ == "__main__":
    markdown_text = """
    # 这是一个标题
    这是普通段落，**加粗**和*斜体*。

    ## 二级标题
    这里是二级标题下的段落。

    ### 无序列表
    - 项目1
    - 项目2

    ### 有序列表
    1. 第一项
    2. 第二项

    ### 链接
    [Google](https://www.google.com)

    ### 图片
    ![示例图片](https://example.com/image.png)

    ### 代码块
    ```python
    print("Hello, World!")
    ```
    """

    html_text = markdown_to_html(markdown_text)
    print(html_text)



<p>    # 这是一个标题</p>
<p>    这是普通段落，<strong>加粗</strong>和<em>斜体</em>。</p>

<p>    ## 二级标题</p>
<p>    这里是二级标题下的段落。</p>

<p>    ### 无序列表</p>
<p>    - 项目1</p>
<p>    - 项目2</p>

<p>    ### 有序列表</p>
<p>    1. 第一项</p>
<p>    2. 第二项</p>

<p>    ### 链接</p>
<p>    <a href="https://www.google.com">Google</a></p>

<p>    ### 图片</p>
<p>    !<a href="https://example.com/image.png">示例图片</a></p>

<p>    ### 代码块</p>
<p>    ``<code>python</p>
<p>    print("Hello, World!")</p>
<p>    </code>``</p>
<p>    </p>


九、邮件轰炸机
你的一个朋友生日快到了，你决定搞个恶作剧——用邮件塞满他的邮箱。
功能描述：首先去注册十个邮箱，然后用这是个邮箱轮流发送随机生成的邮件内容给你的朋友。调整发送的频率，以免被服务器拒绝。

暂未成功

In [ ]:
import smtplib
import time
import random
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from faker import Faker

# 创建Faker实例来生成随机文本
fake = Faker()

# 邮箱和密码列表（假设你已经有了这些邮箱账户的凭据）
email_accounts = [
    ("email1@example.com", "password1"),
    ("email2@example.com", "password2"),
    ("email3@example.com", "password3"),
    ("email4@example.com", "password4"),
    ("email5@example.com", "password5"),
    ("email6@example.com", "password6"),
    ("email7@example.com", "password7"),
    ("email8@example.com", "password8"),
    ("email9@example.com", "password9"),
    ("email10@example.com", "password10")
]

# 收件人邮箱（假设你的朋友的邮箱）
friend_email = "friend@example.com"

# 发送邮件的SMTP服务器（例如Gmail的SMTP服务器）
smtp_server = "smtp.example.com"
smtp_port = 587

# 设置发送间隔（秒）
send_interval = 5  # 每5秒发送一次邮件

def create_random_email_content():
    """生成随机邮件内容"""
    subject = fake.sentence()  # 随机生成一个句子作为邮件主题
    body = fake.text()  # 随机生成一段文本作为邮件正文
    return subject, body

def send_email(from_email, from_password, to_email, subject, body):
    """发送邮件"""
    try:
        # 连接到SMTP服务器
        server = smtplib.SMTP(smtp_server, smtp_port)
        server.starttls()  # 启用TLS加密
        server.login(from_email, from_password)  # 登录SMTP服务器

        # 创建邮件内容
        msg = MIMEMultipart()
        msg['From'] = from_email
        msg['To'] = to_email
        msg['Subject'] = subject
        msg.attach(MIMEText(body, 'plain'))

        # 发送邮件
        server.sendmail(from_email, to_email, msg.as_string())

        print(f"邮件发送成功: {from_email} -> {to_email}")

    except Exception as e:
        print(f"发送邮件失败: {from_email} -> {to_email}, 错误: {e}")
    finally:
        server.quit()

def main():
    email_counter = 0
    while True:
        # 轮流使用邮箱账户发送邮件
        from_email, from_password = email_accounts[email_counter]
        
        # 生成随机邮件内容
        subject, body = create_random_email_content()

        # 发送邮件
        send_email(from_email, from_password, friend_email, subject, body)

        # 更新邮箱计数器，确保轮流使用不同邮箱
        email_counter = (email_counter + 1) % len(email_accounts)

        # 等待一段时间再发送下一封邮件
        time.sleep(send_interval)

if __name__ == "__main__":
    main()


ModuleNotFoundError: No module named 'faker'

十、用邮件控制电脑
你的电脑放在宿舍，这时你在教室上课，你又决定搞个恶作剧吓吓装病请假在宿舍玩游戏的室友。
功能描述：使用邮件发送特定的命令，让电脑上的程序判断发信方是否为控制邮箱，如果是就执行发送过来的命令。通过不同的方式实现尽可能多的功能。


这个效果还不清楚-暂未成功

In [ ]:
# 设置和导入依赖
import imaplib
import email
from email.header import decode_header
import subprocess
import time
import smtplib
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart

# 控制邮箱的配置
control_email = "control@example.com"  # 控制邮箱
control_password = "your_control_email_password"  # 控制邮箱的密码
imap_server = "imap.example.com"  # IMAP服务器
smtp_server = "smtp.example.com"  # SMTP服务器
smtp_port = 587  # SMTP端口
send_email = "send@example.com"  # 用于发送确认邮件的邮箱
send_password = "your_send_email_password"  # 发送确认邮件的邮箱密码

# 定义允许的命令列表
allowed_commands = ["shutdown", "restart", "date", "uptime"]

# 用来发送邮件的函数
def send_confirmation_email(subject, body):
    msg = MIMEMultipart()
    msg['From'] = send_email
    msg['To'] = control_email
    msg['Subject'] = subject
    msg.attach(MIMEText(body, 'plain'))
    
    try:
        with smtplib.SMTP(smtp_server, smtp_port) as server:
            server.starttls()  # 启用TLS加密
            server.login(send_email, send_password)  # 登录
            server.sendmail(send_email, control_email, msg.as_string())  # 发送邮件
            print("确认邮件发送成功")
    except Exception as e:
        print(f"发送邮件失败: {e}")



#2. 连接IMAP服务器并检查新邮件
#使用IMAP协议连接邮箱，检查是否有新邮件。
def check_new_email():
    try:
        # 连接到IMAP服务器
        mail = imaplib.IMAP4_SSL(imap_server)
        mail.login(control_email, control_password)  # 登录控制邮箱
        mail.select('inbox')  # 选择收件箱
        
        # 搜索所有未读邮件
        status, messages = mail.search(None, '(UNSEEN)')
        
        if status != "OK":
            print("没有新邮件")
            return None
        
        # 获取邮件ID
        email_ids = messages[0].split()
        
        # 获取最新的一封邮件
        latest_email_id = email_ids[-1]
        
        # 获取邮件数据
        status, msg_data = mail.fetch(latest_email_id, '(RFC822)')
        
        for response_part in msg_data:
            if isinstance(response_part, tuple):
                # 解析邮件
                msg = email.message_from_bytes(response_part[1])
                subject, encoding = decode_header(msg['Subject'])[0]
                if isinstance(subject, bytes):
                    subject = subject.decode(encoding or 'utf-8')
                from_ = msg.get('From')
                
                print(f"收到来自{from_}的邮件，主题：{subject}")
                
                # 只处理控制邮箱发送的邮件
                if control_email in from_:
                    return subject, msg.get_payload(decode=True).decode('utf-8')  # 返回主题和邮件正文
    except Exception as e:
        print(f"检查邮件时出错: {e}")
    return None


# 3. 执行命令并验证安全性
# 从邮件正文中提取命令并执行。命令列表可以根据需求扩展。需要对命令进行验证，避免执行不安全的命令。
def execute_command(command):
    if command in allowed_commands:
        if command == "shutdown":
            print("执行关机命令")
            subprocess.run(["shutdown", "/s", "/f", "/t", "0"])  # 关机命令
        elif command == "restart":
            print("执行重启命令")
            subprocess.run(["shutdown", "/r", "/f", "/t", "0"])  # 重启命令
        elif command == "date":
            print("执行查看日期命令")
            subprocess.run(["date"])  # 显示当前日期
        elif command == "uptime":
            print("执行查看系统运行时间命令")
            subprocess.run(["uptime"])  # 显示系统运行时间
        else:
            print("命令不在允许的列表中")
    else:
        print("命令不在允许的命令列表中")


#4. 定期检查邮件并执行命令
#现在，你可以将所有功能结合起来，定期检查邮件并执行命令。
def main():
    while True:
        print("检查新邮件...")
        result = check_new_email()
        if result:
            subject, body = result
            print(f"命令: {body}")
            
            # 执行命令
            execute_command(body.strip())
            
            # 发送确认邮件
            send_confirmation_email("命令执行确认", f"收到并执行了命令：{body.strip()}")
        
        # 每隔一定时间检查一次
        time.sleep(60)  # 每60秒检查一次新邮件

if __name__ == "__main__":
    main()


检查新邮件...
检查邮件时出错: [Errno -2] Name or service not known
检查新邮件...
检查邮件时出错: [Errno -2] Name or service not known


KeyboardInterrupt: 